### ISSUE TITLE
Using `.batchLimit(0)` stops the stream observer from getting any events from Nakadi

### ISSUE THREAD

**Issue by dcherepovsky** [zalando-incubator/nakadi-java#125]
In JavaDoc it says:
```scala
/**
 * Maximum number of Events in each batch of the stream. If 0 or unspecified
 * will buffer Events indefinitely and flush on reaching of {@link #batchFlushTimeoutSeconds()}.
 */
```
My configuration:
```scala
val streamConfiguration = new StreamConfiguration()
  .subscriptionId(subscriptionId)
  .batchFlushTimeout(1, TimeUnit.SECONDS)
  .batchLimit(batchLimit) // where batchLimit is 0
  .maxUncommittedEvents(batchLimit) // this line doesn't matter. Tried without it.
```

**Comment by dehora** [comment]
> Using `.batchLimit(0)` stops the stream observer from getting any events from Nakadi 

I think this is happening upstream on the server as the same thing can be seen from curl. Here's a request with a `batch_limit` of `0` with up to 5K uncommitted events:

```curl
curl -v http://localhost:9080/subscriptions/7dd89e8a-65fe-4d45-a982-5088e26d8077/events?batch_limit=0&batch_flush_timeout=3&max_uncommitted_events=5000
```

it will immediately start issuing keepalives:

```bash
{"cursor":{"partition":"0","offset":"BEGIN","event_type":"et_125","cursor_token":"4ffacc83-45df-43ce-8656-a079b2caf867"},"info":{"debug":"Stream started"}}
{"cursor":{"partition":"0","offset":"BEGIN","event_type":"et_125","cursor_token":"04e48714-6c43-4977-a18a-626a8d6fbb95"}}
{"cursor":{"partition":"0","offset":"BEGIN","event_type":"et_125","cursor_token":"4c9c057a-32e2-4f64-95a4-ff0632b9f3bb"}}
{"cursor":{"partition":"0","offset":"BEGIN","event_type":"et_125","cursor_token":"f7a29eb2-4b93-4b9e-91e4-00f394deb9c9"}}
{"cursor":{"partition":"0","offset":"BEGIN","event_type":"et_125","cursor_token":"4e648018-a25d-479f-8275-95b500a07059"}}
```

and never issues events, even as they're being added.

Here's the same subscription request this time without a `batch_limit`:

```bash
curl -v http://localhost:9080/subscriptions/7dd89e8a-65fe-4d45-a982-5088e26d8077/events?&batch_flush_timeout=3&max_uncommitted_events=5000
```

It immediately starts streaming events:

```bash
{"cursor":{"partition":"0","offset":"000000000000000000","event_type":"et_125","cursor_token":"66046851-e007-449c-bcda-adc3dba41d51"},"events":[{"metadata":{"occurred_at":"2017-04-19T16:06:24.447+01:00","eid":"a62569fa-b315-4c03-83c4-86a4b3c5ef82","event_type":"et_125","partition":"0","flow_id":"njc-1492614384470-39044c2bc2b5d59e","received_at":"2017-04-19T15:06:24.549Z","version":"1.0.0"},"data_op":"C","data":{"id":"0:0"},"data_type":"et_125"}],"info":{"debug":"Stream started"}}
...
{"cursor":{"partition":"0","offset":"000000000000000089","event_type":"et_125","cursor_token":"8c2a6522-6c30-4132-8038-daaa2b0524d9"},"events":[{"metadata":{"occurred_at":"2017-04-19T16:24:09.113+01:00","eid":"a9a5595b-2df9-43c3-b611-913c54eb2f86","event_type":"et_125","partition":"0","flow_id":"njc-1492615449113-6857311d9f5dba17","received_at":"2017-04-19T15:24:09.002Z","version":"1.0.0"},"data_op":"C","data":{"id":"@@@:1:1"},"data_type":"et_125"}]}
{"cursor":{"partition":"0","offset":"000000000000000089","event_type":"et_125","cursor_token":"537aca73-ba43-4da0-b500-a3efdbef81b7"}}
```

and will issue events added after joining the stream:

```bash
{"cursor":{"partition":"0","offset":"000000000000000095","event_type":"et_125","cursor_token":"ad8527d6-a8e8-4f59-b3c0-c5354b36411a"}}
{"cursor":{"partition":"0","offset":"000000000000000095","event_type":"et_125","cursor_token":"1be98771-e3cd-4cc2-b6c6-5c5bdec8b30a"}}
{"cursor":{"partition":"0","offset":"000000000000000095","event_type":"et_125","cursor_token":"7491bcb5-e0fc-4501-9a5d-dde080372c00"}}
{"cursor":{"partition":"0","offset":"000000000000000096","event_type":"et_125","cursor_token":"f472a81e-9e28-4302-8a03-cef3f29534aa"},"events":[{"metadata":{"occurred_at":"2017-04-19T16:26:57.456+01:00","eid":"3c79da6a-11b5-4eaa-a10b-ffe27df528bc","event_type":"et_125","partition":"0","flow_id":"njc-1492615617469-43f81c0f2f0f6c92","received_at":"2017-04-19T15:26:57.357Z","version":"1.0.0"},"data_op":"C","data":{"id":"@@@:0:0"},"data_type":"et_125"}]}
{"cursor":{"partition":"0","offset":"000000000000000097","event_type":"et_125","cursor_token":"612c6db1-1b69-4f12-903b-35a724232060"},"events":[{"metadata":{"occurred_at":"2017-04-19T16:26:57.46+01:00","eid":"4017e382-3e34-4253-9984-af31249d892d","event_type":"et_125","partition":"0","flow_id":"njc-1492615617469-43f81c0f2f0f6c92","received_at":"2017-04-19T15:26:57.357Z","version":"1.0.0"},"data_op":"C","data":{"id":"@@@:0:1"},"data_type":"et_125"}]}
```

until the stream is disconnected due to non-checkpointing or the 5K limit is reached.

I'd say this needs to be filed upstream with the Nakadi project itself in case there's a bug or the interaction between parameters isn't clear - the javadoc quoted above is taken more or less verbatim from Nakadi's API definition and the client doesn't anything clever with url parameters.

**Comment by dehora** [comment]
> I'd say this needs to be filed upstream with the Nakadi project itself in case 
> there's a bug or the interaction between parameters isn't clear -

Talking to the project team, batch limit of 0 isn't a useful value and the API definition is wrong here: https://github.com/zalando/nakadi/blob/master/api/nakadi-event-bus-api.yaml#L2218-L2219

**Comment by dcherepovsky** [comment]
OK, will remove it from our code. Hope not to forget about that :)

**Comment by dehora** [comment]
@dcherepovsky there's a pr on the way that will protect against this. Should be available in about 30 minutes.

**Comment by dehora** [comment]
ok, #132's available in 0.7.8.
